df_operator/src/udfs/time_bucket.rs

```diff
@@ -235,89 +232,87 @@ impl Period {
-        let ts_offset = week_offset * week_millis;
-        // TODO(yingwen): Impl sub/divide for Timestamp
-        let week_millis = i64::from(week_millis);
-        let truncated_ts = (ts.as_i64() - i64::from(ts_offset)) / week_millis * week_millis;
-
-        Timestamp::new(truncated_ts)
+    fn truncate_by_date_trunc(
+        array: &DfColumnarValue,
+        granularity: &str,
+    ) -> Result<DfColumnarValue> {
+        let granularity = DfColumnarValue::Scalar(ScalarValue::Utf8(Some(granularity.to_string())));
+        let array = Self::fix_offset(array)?;
+        let trunc_array = date_trunc(&[granularity, array]).context(TruncateTimestamp)?;
+        let result_array = Self::fix_result_offset(&trunc_array)?;
+        Ok(result_array)
     }
 
-    fn truncate_month(ts: Timestamp) -> Timestamp {
-        let offset = FixedOffset::east_opt(DEFAULT_TIMEZONE_OFFSET_SECS).expect("won't panic");
-        // Convert to local time. Won't panic.
-        let datetime = offset.timestamp_millis_opt(ts.as_i64()).unwrap();
-
-        // Truncate month. Won't panic.
-        let truncated_datetime = offset
-            .with_ymd_and_hms(datetime.year(), datetime.month(), 1, 0, 0, 0)
-            .unwrap();
-        let truncated_ts = truncated_datetime.timestamp_millis();
+    fn fix_result_offset(array: &DfColumnarValue) -> Result<DfColumnarValue> {
+        let list = match array {
+            DfColumnarValue::Array(array) => {
+                let array = as_timestamp_nanosecond_array(&array).context(TruncateTimestamp)?;
+                array
+                    .iter()
+                    .map(|ts| {
+                        ts.map(|t| Ok(t - DEFAULT_TIMEZONE_OFFSET_SECS as i64 * 1_000_000_000))
+                            .transpose()
+                    })
+                    .collect::<Result<TimestampNanosecondArray>>()?
+            }
+            _ => return UnsupportedScalar.fail(),
+        };
 
-        Timestamp::new(truncated_ts)
+        Ok(DfColumnarValue::Array(Arc::new(list)))
     }
 
-    fn truncate_year(ts: Timestamp) -> Timestamp {
-        let offset = FixedOffset::east_opt(DEFAULT_TIMEZONE_OFFSET_SECS).expect("won't panic");
-        // Convert to local time. Won't panic.
-        let datetime = offset.timestamp_millis_opt(ts.as_i64()).unwrap();
-
-        // Truncate year. Won't panic.
-        let truncated_datetime = offset
-            .with_ymd_and_hms(datetime.year(), 1, 1, 0, 0, 0)
-            .unwrap();
-        let truncated_ts = truncated_datetime.timestamp_millis();
+    fn fix_offset(array: &DfColumnarValue) -> Result<DfColumnarValue> {
+        let list = match array {
+            DfColumnarValue::Array(array) => {
+                let array = as_timestamp_nanosecond_array(&array).context(TruncateTimestamp)?;
+                array
+                    .iter()
+                    .map(|ts| {
+                        ts.map(|t| Ok(t + DEFAULT_TIMEZONE_OFFSET_SECS as i64 * 1_000_000_000))
```

Comment:
I don't understand here.

Usually first you need to convert timestamp to local datetime, then truncate by week/month/year.

You can refer how datafusion implement similar functions:
https://github.com/apache/arrow-datafusion/blob/8ada7fd1e949b1ff2c9207ce2143bb19157e75cd/datafusion/physical-expr/src/datetime_expressions.rs#L218

Reply:
`date_trunc` is converted to UTC after receiving the timestamp, which subtracts the time zone, This will result in the wrong time zone. what do I think is wrong?😂

Reply:
If `date_trunc` convert UTC datetime, then we can't using it directly...

I will check this carefully later.